traffic_portal/app/src/common/modules/table/topologyCacheGroups/TableSelectTopologyCacheGroupsController.js

```diff
@@ -0,0 +1,196 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+var TableSelectTopologyCacheGroupsController = function(parent, topology, cacheGroups, usedCacheGroupNames, $scope, $uibModal, $uibModalInstance, serverService) {
+
+	let selectedCacheGroups = [],
+		usedCacheGroupCount = 0;
+
+	let addAll = function() {
+		markVisibleCacheGroups(true);
+	};
+
+	let removeAll = function() {
+		markVisibleCacheGroups(false);
+	};
+
+	let markVisibleCacheGroups = function(selected) {
+		let visibleCacheGroupIds = $('#availableCacheGroupsTable tr.cg-row').map(
+			function() {
+				return parseInt($(this).attr('id'));
+			}).get();
+		$scope.cacheGroups = cacheGroups.map(function(cg) {
+			if (visibleCacheGroupIds.includes(cg.id)) {
+				cg['selected'] = selected;
+			}
+			return cg;
+		});
+		updateSelectedCount();
+	};
+
+	let decorateCacheGroups = function() {
+		$scope.cacheGroups = cacheGroups.map(function(cg) {
+			let isUsed = usedCacheGroupNames.find(function(usedCacheGroupName) { return usedCacheGroupName === cg.name });
+			if (isUsed) {
+				cg['selected'] = true;
+				cg['used'] = true;
+				usedCacheGroupCount++;
+			}
+			return cg;
+		});
+	};
+
+	let updateSelectedCount = function() {
+		selectedCacheGroups = $scope.cacheGroups.filter(function(cg) { return cg['selected'] == true && !cg['used'] } );
+		$('div.selected-count').html('<strong><span class="text-success">' + selectedCacheGroups.length + ' selected</span><span> | ' + usedCacheGroupCount + ' already used in topology</span></strong>');
```

Comment:
> `s/\$\(([^\)]+)\)/document\.querySelectorAll\(\1\)/g`

Reply:
Manipulating HTML in a controller like this would ideally just be done with binding and an ng-if - why do it like this?

Reply:
pretty sure i went down the angular approach initially and due to problems resorted to this....and it works :)

Reply:
`ng-if` doesn't work?

Reply:
no, it's because this is an element embedded in the jquery datatable and defined like this:

`"dom": '<"selected-count">frtip',`

it's funky to say the least.

Reply:
Oh right. Yeah, I can see that. But! That doesn't mean we need jQuery!

Besides jQuery + Angular = bad, the other problem with this is direct manipulation of innerHTML (which I assume is what `.html` does, because idk how else it could do it that wouldn't be even worse). That not only causes problems when the hard-coded HTML isn't valid, or when the interpolated values you're using contain characters that manipulate the DOM tree, but it's also actually really slow because it forces the browser to invoke its HTML parser. Plus, because every call to the method uses `.html`, it takes those DOM nodes that already exist and destroys them then re-creates them on every click of a checkbox in that window, instead of just manipulating the text of the existing nodes.

You can avoid both of those performance problems - and jQuery - by doing this:
```diff
diff --git a/traffic_portal/app/src/common/modules/table/topologyCacheGroups/TableSelectTopologyCacheGroupsController.js b/traffic_portal/app/src/common/modules/table/topologyCacheGroups/TableSelectTopologyCacheGroupsController.js
index 3ff080965..c2ab63823 100644
--- a/traffic_portal/app/src/common/modules/table/topologyCacheGroups/TableSelectTopologyCacheGroupsController.js
+++ b/traffic_portal/app/src/common/modules/table/topologyCacheGroups/TableSelectTopologyCacheGroupsController.js
@@ -58,7 +58,30 @@ var TableSelectTopologyCacheGroupsController = function(parent, topology, cacheG
 
 	let updateSelectedCount = function() {
 		selectedCacheGroups = $scope.cacheGroups.filter(function(cg) { return cg['selected'] == true && !cg['used'] } );
-		$('div.selected-count').html('<strong><span class="text-success">' + selectedCacheGroups.length + ' selected</span><span> | ' + usedCacheGroupCount + ' already used in topology</span></strong>');
+
+		const selectedText = String(selectedCacheGroups.length) + " selected";
+		const usedText = " | " + String(usedCacheGroupCount) + " already used in topology";
+
+		let selectedElement = document.getElementById("cache-groups-selected-count");
+		if (selectedElement === null) {
+			selectedElement = document.createElement("STRONG");
+			selectedElement.id = "cache-groups-selected-count";
+
+			const selected = document.createElement("SPAN");
+			selected.className = "text-success";
+			selected.textContent = selectedText;
+
+			const used = document.createElement("SPAN");
+			used.textContent = usedText;
+			selectedElement.appendChild(selected);
+			selectedElement.appendChild(used);
+
+			document.getElementsByClassName("selected-count")[0].appendChild(selectedElement);
+			return;
+		}
+
+		selectedElement.firstChild.textContent = selectedText;
+		selectedElement.lastChild.textContent = usedText;
 	};
 
 	$scope.parent = parent;
```
Which is certainly more verbose, but much more performant. And doesn't use jQuery.